### PR TITLE
flatten uuid strings

### PR DESCRIPTION
**Pull request by aardgoose**
Workaround for #13069 issues

### CODE REVIEW

**Comment by mrdoob** [comment]
We may want to run this test before and after:

https://github.com/mrdoob/three.js/issues/12432#issuecomment-338059896

These are the numbers that I got the last time:

https://github.com/mrdoob/three.js/issues/12432#issuecomment-342952905

**Comment by mrdoob** [comment]
```js
var _Math = {

	generateUUID: ( function () {

		// http://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid-in-javascript/21963136#21963136

		var lut = [];

		for ( var i = 0; i < 256; i ++ ) {

			lut[ i ] = ( i < 16 ? '0' : '' ) + ( i ).toString( 16 ).toUpperCase();

		}

		return function generateUUID() {

			var d0 = Math.random() * 0xffffffff | 0;
			var d1 = Math.random() * 0xffffffff | 0;
			var d2 = Math.random() * 0xffffffff | 0;
			var d3 = Math.random() * 0xffffffff | 0;
			return lut[ d0 & 0xff ] + lut[ d0 >> 8 & 0xff ] + lut[ d0 >> 16 & 0xff ] + lut[ d0 >> 24 & 0xff ] + '-' +
				lut[ d1 & 0xff ] + lut[ d1 >> 8 & 0xff ] + '-' + lut[ d1 >> 16 & 0x0f | 0x40 ] + lut[ d1 >> 24 & 0xff ] + '-' +
				lut[ d2 & 0x3f | 0x80 ] + lut[ d2 >> 8 & 0xff ] + '-' + lut[ d2 >> 16 & 0xff ] + lut[ d2 >> 24 & 0xff ] +
				lut[ d3 & 0xff ] + lut[ d3 >> 8 & 0xff ] + lut[ d3 >> 16 & 0xff ] + lut[ d3 >> 24 & 0xff ];

		};

	} )(),

	generateUUID2: ( function () {

		// http://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid-in-javascript/21963136#21963136

		var lut = [];

		for ( var i = 0; i < 256; i ++ ) {

			lut[ i ] = ( i < 16 ? '0' : '' ) + ( i ).toString( 16 );

		}

		return function generateUUID() {

			var d0 = Math.random() * 0xffffffff | 0;
			var d1 = Math.random() * 0xffffffff | 0;
			var d2 = Math.random() * 0xffffffff | 0;
			var d3 = Math.random() * 0xffffffff | 0;

			return String( lut[ d0 & 0xff ] + lut[ d0 >> 8 & 0xff ] + lut[ d0 >> 16 & 0xff ] + lut[ d0 >> 24 & 0xff ] + '-' +
				lut[ d1 & 0xff ] + lut[ d1 >> 8 & 0xff ] + '-' + lut[ d1 >> 16 & 0x0f | 0x40 ] + lut[ d1 >> 24 & 0xff ] + '-' +
				lut[ d2 & 0x3f | 0x80 ] + lut[ d2 >> 8 & 0xff ] + '-' + lut[ d2 >> 16 & 0xff ] + lut[ d2 >> 24 & 0xff ] +
				lut[ d3 & 0xff ] + lut[ d3 >> 8 & 0xff ] + lut[ d3 >> 16 & 0xff ] + lut[ d3 >> 24 & 0xff ] ).toUpperCase();

		};

	} )()

};

function run( func ) {

	var loop = 0x100000;

	var startTime = performance.now();

	for ( var i = 0; i < loop; i ++ ) {

		func();

	}

	var endTime = performance.now();

	console.log( ( endTime - startTime ).toFixed( 2 ) + 'ms, loop=' + loop );

}

run(_Math.generateUUID);
run(_Math.generateUUID);
run(_Math.generateUUID2);
run(_Math.generateUUID2);
```

**Comment by mrdoob** [comment]
Goes from ~350ms to ~740ms here... 😕

**Comment by makc** [comment]
why does 3js use uuids at all? can't it just use counters?

**Comment by mrdoob** [comment]
> why does 3js use uuids at all? can't it just use counters?

Mainly for serialising/deserialising.

**Comment by makc** [comment]
> deserialising

ok, and why can't you do geometries[123] instead of geometries['XXX-YYY-SOMETHING'] while deserialising?

**Comment by mrdoob** [comment]
I'm wondering if uppercasing the uuid is worth it though. As far as I understand, uppercase uuid consumes 10x more memory than lowercase?

**Comment by mrdoob** [comment]
> ok, and why can't you do geometries[123] instead of geometries['XXX-YYY-SOMETHING'] while deserialising?

Lets not derail this thread.

**Comment by makc** [comment]
in the context of #13069 .toLowerCase() would probably have the same effect

**Comment by aardgoose** [comment]
the performance hit is why i looked at removing the requirement for the uuid's with the lazy_uuid PR. 

The WeakMap.get( object ) is probably a simpler lookup just on the object itself rather than having to get an object property as the uuid lookup in WebGLProperties & Attributes so that is probably a gain too.

@mrdoob the case of the uuid doesn't matter.  It just gets the VM to create a more compact copy as a side effect - hence the hacky nature.

**Comment by mrdoob** [comment]
How about this then:

```js
( function () {

		// http://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid-in-javascript/21963136#21963136

		var lut = [];

		for ( var i = 0; i < 256; i ++ ) {

			lut[ i ] = ( i < 16 ? '0' : '' ) + ( i ).toString( 16 ).toUpperCase();

		}

		return function generateUUID() {

			var d0 = Math.random() * 0xffffffff | 0;
			var d1 = Math.random() * 0xffffffff | 0;
			var d2 = Math.random() * 0xffffffff | 0;
			var d3 = Math.random() * 0xffffffff | 0;

			return String( lut[ d0 & 0xff ] + lut[ d0 >> 8 & 0xff ] + lut[ d0 >> 16 & 0xff ] + lut[ d0 >> 24 & 0xff ] + '-' +
				lut[ d1 & 0xff ] + lut[ d1 >> 8 & 0xff ] + '-' + lut[ d1 >> 16 & 0x0f | 0x40 ] + lut[ d1 >> 24 & 0xff ] + '-' +
				lut[ d2 & 0x3f | 0x80 ] + lut[ d2 >> 8 & 0xff ] + '-' + lut[ d2 >> 16 & 0xff ] + lut[ d2 >> 24 & 0xff ] +
				lut[ d3 & 0xff ] + lut[ d3 >> 8 & 0xff ] + lut[ d3 >> 16 & 0xff ] + lut[ d3 >> 24 & 0xff ] );

		};

	} )()
```

No performance changes with that code here.

**Comment by mrdoob** [comment]
For testing...

```js
var _Math = {

	generateUUID: ( function () {

		// http://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid-in-javascript/21963136#21963136

		var lut = [];

		for ( var i = 0; i < 256; i ++ ) {

			lut[ i ] = ( i < 16 ? '0' : '' ) + ( i ).toString( 16 ).toUpperCase();

		}

		return function generateUUID() {

			var d0 = Math.random() * 0xffffffff | 0;
			var d1 = Math.random() * 0xffffffff | 0;
			var d2 = Math.random() * 0xffffffff | 0;
			var d3 = Math.random() * 0xffffffff | 0;
			return lut[ d0 & 0xff ] + lut[ d0 >> 8 & 0xff ] + lut[ d0 >> 16 & 0xff ] + lut[ d0 >> 24 & 0xff ] + '-' +
				lut[ d1 & 0xff ] + lut[ d1 >> 8 & 0xff ] + '-' + lut[ d1 >> 16 & 0x0f | 0x40 ] + lut[ d1 >> 24 & 0xff ] + '-' +
				lut[ d2 & 0x3f | 0x80 ] + lut[ d2 >> 8 & 0xff ] + '-' + lut[ d2 >> 16 & 0xff ] + lut[ d2 >> 24 & 0xff ] +
				lut[ d3 & 0xff ] + lut[ d3 >> 8 & 0xff ] + lut[ d3 >> 16 & 0xff ] + lut[ d3 >> 24 & 0xff ];

		};

	} )(),

	generateUUID2: ( function () {

		// http://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid-in-javascript/21963136#21963136

		var lut = [];

		for ( var i = 0; i < 256; i ++ ) {

			lut[ i ] = ( i < 16 ? '0' : '' ) + ( i ).toString( 16 ).toUpperCase();

		}

		return function generateUUID() {

			var d0 = Math.random() * 0xffffffff | 0;
			var d1 = Math.random() * 0xffffffff | 0;
			var d2 = Math.random() * 0xffffffff | 0;
			var d3 = Math.random() * 0xffffffff | 0;

			return String(
        lut[ d0 & 0xff ] + lut[ d0 >> 8 & 0xff ] + lut[ d0 >> 16 & 0xff ] + lut[ d0 >> 24 & 0xff ] + '-' +
				lut[ d1 & 0xff ] + lut[ d1 >> 8 & 0xff ] + '-' + lut[ d1 >> 16 & 0x0f | 0x40 ] + lut[ d1 >> 24 & 0xff ] + '-' +
				lut[ d2 & 0x3f | 0x80 ] + lut[ d2 >> 8 & 0xff ] + '-' + lut[ d2 >> 16 & 0xff ] + lut[ d2 >> 24 & 0xff ] +
				lut[ d3 & 0xff ] + lut[ d3 >> 8 & 0xff ] + lut[ d3 >> 16 & 0xff ] + lut[ d3 >> 24 & 0xff ]
      );

		};

	} )()

};

function run( func ) {

	var loop = 0x100000;

	var startTime = performance.now();

	for ( var i = 0; i < loop; i ++ ) {

		func();

	}

	var endTime = performance.now();

	console.log( ( endTime - startTime ).toFixed( 2 ) + 'ms, loop=' + loop );

}

run(_Math.generateUUID);
run(_Math.generateUUID);
run(_Math.generateUUID2);
run(_Math.generateUUID2);
```

**Comment by aardgoose** [comment]
Do a heap dump in Chrome and look at the (concatenated string)
uuid
 section.  You can see the way it stores the strings in lots of small sections. 

 I think just String() alone just produces a reference to the bloated version hence no performance change and not reduction of memory.

**Comment by makc** [comment]
out of interest just tried to do this with Uint8Array and TextDecoder, and [it's only 15 times slower](https://jsfiddle.net/ctengc79/) :D then I thought if I inline setHex, I can gain something, but profiler says -23% max
screen shot 2018-01-11 at 23 05 30

compared to this, toUpperCase x2 perfrmance drop does not look that bad at all.

**Comment by takahirox** [comment]
BTW, `Array.join( '' )` used in older `generateUUID()` seems to allocate one consecutive memory space.  You can try https://github.com/mrdoob/three.js/issues/13069#issuecomment-356602348 by replacing `r89` with `r87`.

**Comment by takahirox** [comment]
Number tech is interesting...
But sometimes it doesn't work well and we need to carefully use it?

I tried four patterns with the following code on my Windows10 + Chrome.

    <script src="https://rawgit.com/mrdoob/three.js/r89/build/three.js"></script>
    <script>
    var num = 0x10000;

    function flatstr( s ) {

        Number( s );
        return s;

    }

    var array = Array( num );

    for ( var i = 0; i < num; i ++ ) {

        // 1.
        //array[ i ] = THREE.Math.generateUUID();

        // 2.
        //array[ i ] = THREE.Math.generateUUID().toUpperCase();

        // 3.
        //array[ i ] = flatstr( THREE.Math.generateUUID() );

        // 4.
        //var uuid = THREE.Math.generateUUID();
        //Number( uuid );
        //array[ i ] = uuid;

    }
    </script>

| | total heap snapshot size | concatenated string count | shallow size | retained size |
|----|----|----|----|----|
| 1. | 42.3MB | 921,259 | 36,850,360 | 39,507,080 |
| 2. | 8.8MB | 3,755 | 150,200 | 185,480 |
| 3. | 38.2MB | 809,262 | 32,370,480 | 34,776,728 |
| 4. | 10.7MB | 52,893 | 2,115,720 | 5,295,832 |

I expected 3.'s and 4.'s result are same as the 2.'s result but seems like a lot of concatenated strings are left in heap.

**Comment by aardgoose** [comment]
I was half expecting this to be a recent issue, since I hadn't seen the problem before when looking at heap size in general.

I would check what you get out from Number( uuid ). console.log( Number ( new THREE.Object3D().uuid ) );   produces NaN.

**Comment by takahirox** [comment]
Probably what `Number( uuid )` produces doesn't matter in this tech if I'm right.

**Comment by makc** [comment]
@takahirox and from 42 vs 38 mb it does not seem it produces anything at all

**Comment by takahirox** [comment]
3 works now by replacing `flatstr` with 

    function flatstr( s ) {

        return isNaN( Number( s ) ) ? s :  '';

    }

Probably we need to let `Number( s )` not be removed by JS engine optimization.

Bad news, `Number( s )` tech isn't fast. A bit slower than `.toUpperCase()`. I speculate String::Flatten method itself isn't fast regardless how we trigger.

**Comment by takahirox** [comment]
Would you reopen this because of https://github.com/mrdoob/three.js/issues/13069#issuecomment-373177567?

File: src/math/Math.js [hunk 1]
```diff
@@ -26,10 +26,11 @@ var _Math = {
 			var d1 = Math.random() * 0xffffffff | 0;
 			var d2 = Math.random() * 0xffffffff | 0;
 			var d3 = Math.random() * 0xffffffff | 0;
-			return lut[ d0 & 0xff ] + lut[ d0 >> 8 & 0xff ] + lut[ d0 >> 16 & 0xff ] + lut[ d0 >> 24 & 0xff ] + '-' +
+
+			return String( lut[ d0 & 0xff ] + lut[ d0 >> 8 & 0xff ] + lut[ d0 >> 16 & 0xff ] + lut[ d0 >> 24 & 0xff ] + '-' +
```
Comment (takahirox) [inline src/math/Math.js]: I prefer removing "String", not strong preference tho.

**Comment by mrdoob** [comment]
I can't reopen this PR because @aardgoose deleted the branch.

**Comment by takahirox** [comment]
Oops, then I make a new one.

**Pull request closed without merging**